cpp/src/decisiontree/decisiontree_impl.cuh

```diff
@@ -75,6 +77,47 @@ void print_node(const std::string &prefix,
   }
 }
 
+template <typename T>
+std::string to_string_high_precision(T x) {
+  static_assert(std::is_floating_point<T>::value || std::is_integral<T>::value,
+                "T must be float, double, or integer");
+  std::ostringstream oss;
+  oss.imbue(std::locale::classic());  // use C locale
+  if (std::is_floating_point<T>::value) {
+    oss << std::setprecision(std::numeric_limits<T>::max_digits10) << x;
```

Comment:
Does this guarantee no loss of precision? Maybe you have a test for this later.

Reply:
https://en.cppreference.com/w/cpp/types/numeric_limits/max_digits10

> The value of `std::numeric_limits<T>::max_digits10` is the number of base-10 digits that are necessary to uniquely represent all distinct values of the type T, such as necessary for serialization/deserialization to text. This constant is meaningful for all floating-point types.